sys/posix/select/posix_select.c

```diff
@@ -0,0 +1,152 @@
+            return 0;
+        }
+        t -= offset;
+        if (t > UINT32_MAX) {
+            errno = EINVAL;
+            /* don't have timer set yet so go to end */
+            return -1;
+        }
+        else {
+            xtimer_set_timeout_flag(timeout_timer, (uint32_t)t);
+        }
+    }
+    return 0;
+}
+
+int select(int nfds, fd_set *readfds, fd_set *writefds, fd_set *errorfds,
+           struct timeval *timeout)
+{
+    uint32_t start_time = xtimer_now_usec();
+    fd_set ret_readfds;
+    xtimer_t timeout_timer;
+    int fds_set = 0;
+    bool wait = true;
+
+    FD_ZERO(&ret_readfds);
+    /* TODO ignored writefds and errorfds for now since there is no point for
+     * them with sockets */
+    if ((nfds >= FD_SETSIZE) || ((unsigned)nfds >= VFS_MAX_OPEN_FILES)) {
+        errno = EINVAL;
+        fds_set = -1;
+        goto end;
+    }
+    for (int i = 0; i < nfds; i++) {
+        if ((readfds != NULL) && FD_ISSET(i, readfds)) {
+            if (!posix_socket_is(i)) {
+                errno = EBADF;
+                fds_set = -1;
+                goto end;
+            }
+            if (posix_socket_avail(i) > 0) {
+                FD_SET(i, &ret_readfds);
+                fds_set++;
+                wait = false;
+            }
+            else {
+                posix_socket_select(i);
+            }
+        }
+        if ((writefds != NULL) && FD_ISSET(i, writefds) &&
+            !posix_socket_is(i)) {
+            errno = EBADF;
+            fds_set = -1;
+            goto end;
+        }
+        if ((errorfds != NULL) && FD_ISSET(i, errorfds) &&
+            !posix_socket_is(i)) {
+            errno = EBADF;
+            fds_set = -1;
+            goto end;
+        }
+    }
+    while (wait) {
+        if (_set_timeout(&timeout_timer, timeout,
+                         xtimer_now_usec() - start_time, &wait) < 0) {
+            /* timeout setting failed so not set, so no need to go to end */
+            return -1;
+        }
+        if (!wait) {
+            /* timeout was reached, timeout was not set, so no need to
+             * go to end */
+            errno = EINTR;
+            return -1;
+        }
+        thread_flags_t tflags = thread_flags_wait_any(POSIX_SELECT_THREAD_FLAG |
+                                                      THREAD_FLAG_TIMEOUT);
+        if (tflags & POSIX_SELECT_THREAD_FLAG) {
+            for (int i = 0; i < nfds; i++) {
+                if (FD_ISSET(i, readfds)) {
+                    if (posix_socket_avail(i) > 0) {
+                        FD_SET(i, &ret_readfds);
+                        fds_set++;
+                        wait = false;
+                    }
+                }
+            }
+        }
+        else if (tflags & THREAD_FLAG_TIMEOUT) {
+            errno = EINTR;
+            /* timeout fired, so no need to go to end */
+            return -1;
+        }
+    }
+    *readfds = ret_readfds;
+end:
+    xtimer_remove(&timeout_timer);
```

Comment:
I am usually for reducing the amount of `return` points in functions. But here all the usages of `end` do not actually need to remove the timer right?  It could be moved before or just return in the other points instead of jumping

Reply:
Right this is an artifact from when the timer was initialized early. Will fix. Should I squash this immediately?

Reply:
Yes

Reply:
Done